# PROFUNGIS

Processing of Fungal ITS Sequences
This pipeline has been created to download SRA studies of fungal ITS markers that will be used to construct a unique set of Z(ero radiance) OTUS to analyse fungal diversity.

## Installing PROFUNGIS
Install before usage:
  - Snakemake
  - BLAST+
  - usearch v11
    - rename the executable file 'usearch11' and place in /deps/ folder

To check if all dependencies are placed correctly, execute the checkDependencies.sh script found in this repo.

## The PROFUNGIS workflow
PROFUNGIS is divided into three main steps: 
- preprocessing
- denoising
- postprocessing

### overview of workflow steps

- **preprocessing**
  
In the preprocessing step, the reads corresponding to the provided SRA run id(s) will be downloaded into the ./sample/ directory. Next, the presence of the provided primers in the primer datafile is checked. 
If one or both of the primers are missing, the user is asked to provide the sequence and the part of ITS it amplifies.
Based on the collection of the user-provided parameters, a configuration file is created (config.yml). This file is required by the denoising part of the pipeline.

- **Denoising**
  
The denoising pipeline follows a mostly basic runthrough of ZOTU
construction. First, the provided primers are cut from both ends of the
reads. If the reads are generated by a paired-end Illumina library, the 
**DecideMerger** script is called. This step is skipped in pyrosequencing
platforms. The DecideMerger script decides if the paired-end reads
should be merged, or that only the forward reads should be used. If
the library has been sequenced using a pyrosequencing platform
(iontorrent or 454), and the amplicon is a partial ITS sequence
(ITS1 or ITS2), the reads are truncated to 250 bp, to ensure
same treatment. Whether or not the amplicon is a partial ITS
sequences is determined based on the user input primer names.
For every primer, its location relative to ITS is added.

Next, the average Estimated Error (EE) is used using the cor-
responding USEARCH function. If the user has provided the argument to use a fixed EE, this is used to filter. If not, mean EE
is used to filter the reads. The next step is dereplication of the
reads. After that, singletons are removed and the denoising step
is initiated. This step used the UNOISE3 algorithm to select
biological sequences without clustering based on a fixed identity
percentage. For more information on this step, see section 6.3.
After this step, the reads are remapped to the ZOTUS and a
ZOTU table containing the number of mapped reads is created.

- **postprocessing**
  
The postprocessing step of the pipeline involves an abundance
filter and a contamination filter. First, the pipeline will use the
ZOTU table to remove all sequences that have less than 0.1% of
the reads in that sample. Next, the contamination filter will use
BLAST to align the ZOTUS to the UNITE database. This is
done to 1) remove putative ZOTUS that are not somewhat sim-
ilar to known ITS sequences and 2) to remove putative ZOTUS
that are not asociated with a fungus. If a ZOTU has a hit with
more than 70% identity and the hit is a fungus, it is retained. If
not, it is removed.

## How to run PROFUNGIS

1. check dependencies
   
```shell
bash checkDependencies . sh
```
Continue if there are no errors. If everything is in place, you can start using PROFUNGIS pipeline. 

2. Run PROFUNGIS pipeline. In order to start, please launch startPROFUNGIS.py script. Please make sure you provide the correct arguments.

** List of mandatory parameters **

the mandatory arguments are:

| Name | acronym used for pipeline | expected input | definition |
| ---- | ---- | ----- | ----- |
| -forward | -f | primer name [String] | label of forward primer to refer to |
| -reverse | -r | primer name [String] | label of reverse primer to refer to |
| -platform | -p | platform name [illumina|454|iontorrent] | HTS plaform used for generating the read sequence file |
| -SRA | -s | one SRA id [String] | SRA sequence read ID (aka SRR) |
| -multirun | -m | file containg SRA ids | a list of SRRs provided in a .txt file |

** run the script** 

In order to run the pipeline, the mandatory arguments are required. This is for running via commandline. See below how to run the scripts in a docker environment. 
The list follows the order of arguments to pass to PROFUNGIS. 
Here we provide an example of running startPROFUNGIS.py with SRA read ID : SRR2002283. This ID is a public SRR obtained from NCBI SRA. Please note that to run startPROFUNGIS.py, you do not need to download the raw sequence files from SRA, this pipeline will do it for you by providing the reference ID.

```shell
python3 startPROFUNGIS . py --forward ExampleFwd -- reverse ExampleRev --platform illumina --SRA SRR2002283
```
The pipeline will read the primer names provided for the forward and reverse primers. These sequences are provided in the primer dataset (primers.py). If the script does not find them, the primers will be generated and saved for you. This is the case if you run for the first time this script. This means that ExamplePwd and ExampleRev are not in the primer file, thus the script will request for you to add them.
For running the first time, you can use and add the following primers:

```shell
CTTGGTCATTTAGAGGAAGTA 
```
for the forward primer. You will need to provide a primer label for this sequence, so you can refer to when running the scripts. In this case we can use the example label: ExampleFwd. This primer belongs to the ITS1 barcode is used.

```shell
GCTGCGTTCTTCATCGATG 
```
for the reverse primer, labelled ExampleRev. This primer is used for ITS1 barcode of ITS.

These primers will be added to the datafile and you can refer to them with *ExampleFwd* and *ExampleRev*. 



